File: story/resources/spark-streaming-pr/cluster/transformation/src/main/scala/aaa/abc/dd/spark_streaming_pr/cluster/BatchService.scala
package aaa.abc.dd.spark_streaming_pr.cluster

import aaa.abc.dd.spark_streaming_pr.cluster.Batch._
import org.apache.spark.ml.regression.{LinearRegression, LinearRegressionModel}
import org.apache.spark.sql.{DataFrame, Row}

// https://spark.apache.org/docs/latest/ml-classification-regression.html#linear-regression
// https://spark.apache.org/docs/latest/ml-pipeline.html

class BatchService extends Train with Predict {
  var service: LinearRegression = _
  var model: LinearRegressionModel = _

  override def train(ds: DataFrame): Unit = {
    service = new LinearRegression().setMaxIter(10).setRegParam(0.01)
    model = service.fit(ds)
  }

  override def predict(ds: DataFrame): Unit = {
    val r = model.transform(ds)
    r.select("label", "prediction")
      .collect()
      .foreach { case Row(label: Double, prediction: Double) =>
        // scalastyle:off println
        println(s"($label) -> prediction=$prediction")
        // scalastyle:on println
      }
  }
}

File: story/resources/spark_streaming_pp/spark_streaming_pp/batch_service.py
from pyspark.ml.regression import LinearRegression


class BatchService:
    def __init__(self):
        self.service = LinearRegression(maxIter=10, regParam=0.01)
        self.model = None

    def train(self, ds):
        self.model = self.service.fit(ds)

    def predict(self, ds):
        prediction = self.model.transform(ds)
        result = prediction.select("label", "prediction").collect()
        for row in result:
            print(f"{row.label}, {row.prediction}")
